### PR TITLE
Addition of threshold parameter in early stopping

### PR DESCRIPTION
**Pull request by MrinalTyagi**
This will allow the user to add a threshold value for the quantity they want like accuracy or loss and during training if the value reaches less than or greater than value of threshold(depending on mode parameter), training will stop. Helps in reducing overfitting.

### CODE REVIEW

File: keras/callbacks.py [hunk 1]
```diff
@@ -1817,6 +1822,12 @@ def on_epoch_end(self, epoch, logs=None):
     current = self.get_monitor_value(logs)
     if current is None:
       return
+    if self.mode == "min" and self.threshold != None and current < self.threshold:
+      self.model.stop_training = True
+      return
```
Comment (old-school-kid) [inline keras/callbacks.py:1827]: You don't have to return from the function. If the current epoch was best and `restore_best_weights= True` then the weights wont ever be restored which is done in the code following this line.

Comment (MrinalTyagi) [reply keras/callbacks.py]: Done

**Comment by MrinalTyagi** [comment]
@old-school-kid Updated with requested changes. Please have a look and update accordingly if any more changes are required.

**Comment by fchollet** [comment]
Thanks for the PR. I actually don't understand what this is for, can you explain the use case like I'm 5?

**Comment by MrinalTyagi** [comment]
> Thanks for the PR. I actually don't understand what this is for, can you explain the use case like I'm 5?

In the case of overfitting of a model, the value of loss as well as other metrics might increase. In case if validation set is not taken, it might be difficult to stop training and detect overfitting which is usually the case when someone has a small dataset and it's difficult to split the dataset into train test validation. Hence this will make sure that model stops training after a certain value of the metrics specified to monitor so that the model does not overfit in case of no validation set as well as in case of validation set it will be helpful. Hopefully, I am able to explain what you were expecting. :)

**Comment by old-school-kid** [comment]
1. Incase a validation data is provided, isn't the gap between validation and training accuracy/loss sufficient enough data to detect overfitting?
2. Incase a validation data isn't provided can't a low accuracy like 80% also mean overfitting?

**Comment by MrinalTyagi** [comment]
> 1. Incase a validation data is provided, isn't the gap between validation and training accuracy/loss sufficient enough data to detect overfitting?
> 2. Incase a validation data isn't provided can't a low accuracy like 80% also mean overfitting?

I think in case of no validation data it will be helpful as yes even 80 percent accuracy means overfitting but setting a specific threshold will allow stopping training after a specific threshold thereby eliminating some of that issue.

**Comment by fchollet** [comment]
Thanks for the explanation! In general it is fairly confusing to have a single callback do multiple things, and it is strongly encouraged to use the current EarlyStopping callback with a validation set. I would suggest using your own custom callback for your feature, and keeping it doing only one thing. We will not merge this.
